services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java

```diff
@@ -790,6 +794,9 @@ private void updateAuthorizationSettings(ClientRepresentation rep) {
      */
     private UserSessionRepresentation toUserSessionRepresentation(final UserSessionModel userSession) {
         UserSessionRepresentation rep = ModelToRepresentation.toRepresentation(userSession);
+        if (rep == null) {
```

Comment:
If you are filtering at the source (`InfinispanUserSessionProvider`) you shouldn't worry about sessions not bound with a user.

Reply:
I don't think that is a good idea. The interface of `UserLookupProvider.getUserById(RealmModel realm, String id) `states that `null` is a valid return value. That means, every call to this function might produce `null` as result. Unless `InfinispanUserSessionProvider` returns the UserModel somehow in its data structure a null check of the caller is required. Especially in such constructs as `session.users().getUserById(getRealm(), entity.getUserId());`. This is used quite often. There you never know how a the interface `UserLookupProvider` is implemented and which constrains are attached to the component.

Reply:
As mentioned before, the places where the session is being used expect valid sessions, bound to a user.

If that is true. There is no reason to duplicate code (filtering by user not null) everywhere.

@mposolda Could confirm if my assumption is correct ....

Reply:
And as I mentioned before, `session.users().getUserById(realm, entity.getUser())` may return `null` any time. As long as `org.keycloak.models.sessions.infinispan.UserSessionAdapter.getUser()` still calls `session.users().getUserById(realm, entity.getUser())` there is a chance that `getUser()` returns `null`. And this must be checked, regardless if it has been checked before. This function is called at `ModelToRepresentation.toRepresentation(UserSessionModel session)`.

Also, in case the `InfinispanUserSessionProvider` checks the user for each element in the stream, this means that every time a lookup of the user is made (using a federated provider; LDAP, or a custom one, like in our case). And then again when the result is used, because `getUser()` makes the call again!

This results in a lot of round-trips! Even from a performance point of view, I see this a bad design.

In my opinion the first thing to be done is to define the expectations of the interfaces. Is there a design document somewhere?

Please recall, I started this PR to fix some NPEs. This is still my goal. I do not want to redesign this whole thing. But with 150k+ concurrent sessions in our system I do not see that our infrastructure is pretty happy doubling the number of requests just because the check for a valid user is done deep down the stack. (Again, it needs to be done at every call of `session.users().getUserById(...)`!)

Reply:
Nevertheless and regardless of my concerns, if you insist I can rewrite this PR to be closer to what I assume is your preferred solution. I need the NPEs fixed. Please let me know your decision.

Reply:
I had the wrong assumption that `UserSessionAdapter` (infinispan)  was leveraging the user cache. I see the problem when resolving the user from the session.

Looks like there are other things to evaluate here but unrelated to this PR. For instance, currently we are still making multiple calls to the underlying storage if `UserSessionAdapter.getUser()` is called multiple times during the request lifecycle.

Changes LGTM but we need some tests. Are you familiar with our test suite or do you want some help?

Reply:
I see both points.

I agree that currently when `UserSessionProvider` returns some sessions, it doesn't check if user connected with the userSession exists. So we assume that it is responsibility of the caller to check if `userSession.getUser()` is null (and `userSession.getUser()` calls lookup of user in `UserProvider`).

If we change this behaviour, that `UserSessionProvider` returns only sessions, where user was proved to exists. This has big advantage, that we won't need to check `if (userSession.getUser() == null)` on various places inside codebase, which would be very great. Basically all sessions returned by userSessionProvider would be guaranteed to be valid sessions with valid user.

On the other hand, I am not sure about performance side-effect of this? As `UserSessionProvider` will need to verify existence of the user before the userSession is returned from `UserSessionProvider`. However performance side-effects may not be that big because we use Stream API?

I think it may be good to involve store team for this discussion.

@hmlnarik any opinion?

Should we just go with the changes in this PR (so verify nulls on the caller) and then consider whether we want to improve implementation of userSessionProvider itself? In that case, I think we should do it for both old and new store.

Reply:
In case the user is checked in the `UserSessionProvider` then the session object should be changed to already include the user and do not look it up again.

The performance side-effect comes into play when (like in our case) the user federation provider is configured with a NOCACHE policy. Then every call to `session.users().getUserById(...)` ends up at the provider and produces network traffic and so on. Therefore the user lookup should be reduced to a minimum. And the Stream API does no reduce the amount of user lookups at all.

Reply:
> Changes LGTM but we need some tests. Are you familiar with our test suite or do you want some help?

Unfortunately, I am not familiar with the test suite. I will have a deeper look next week.

Reply:
Hello all, this is a tough decision to make. 

We had a call with @hmlnarik around this and we believe this needs to be considered case by case. 

In general, moving such integrity checks to the storage makes sense. It results in cleaner logical layer code as well as performance gain as we stop the processing for invalid entities sooner in the process chain. However, the latter doesn't need to be true in all cases. For example, if we will check all parent groups when obtaining a group from storage we can end up with a huge performance impact. 

For this specific use-case, I would lean towards the suggestion from @pedroigor and @mposolda. Sessions for not existing users are invalid and in a perfect world, they would be invisible to Keycloak (for example, if everything would be stored in one database, the sessions would be removed as part of some cascade removal). In cases such as this, where the user was removed from an external system we need to do our best to find this out as soon as possible and therefore I would add this check to the storage. Similarly, we are doing this for client sessions:
https://github.com/keycloak/keycloak/blob/4f8d67c9fcbc1a50976128584369d9a1a9077ca2/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/UserSessionAdapter.java#L110-L126

Note: to avoid loading the user again and again from the LDAP, we can change the `getUser` getter to something like:
```java
public UserModel getUser() {
    if (this.user == null) {
        this.user = session.users().getUserById(realm, entity.getUser());
    }

    return this.user;
}

```
CC: @hmlnarik

Reply:
@mhajas + @hmlnarik: I will change the PR and come up with a solution closer to your suggestion. Stay tuned.

Reply:
Thank you @mkrueger92 !